File: EstruturasFlexiveis/ListaSimplesmenteEncadeada/Java/LinkedList.java
package EstruturasFlexiveis.ListaSimplesmenteEncadeada.Java;

public class LinkedList<T> {
    //=====PRIVATE=====//
    private int length = 0;
    private Node<T> first;
    private Node<T> last;

    //=====CONSTRUCTOR=====//
    public LinkedList() {
        this.setFirst(null);
        this.setLast(null);
    }

    @SuppressWarnings("unchecked")
    public LinkedList(T... arr) {
        for(T obj : arr)
            add(obj);
    }

    //=====GET=====//
    //-----PRIVATE-----//
    private Node<T> getFirst() {
        return this.first;
    }

    private Node<T> getLast() {
        return this.last;
    }

    private Node<T> node(int idx) {
        if(idx < 0 || idx > this.size()) throw new IndexOutOfBoundsException();

        Node<T> curr;
        if(idx == this.size() - 1) curr = this.getLast();
        else {
            curr = this.getFirst();
            int i = 0;  
            while(i < idx) {
                curr = curr.getNext();
                i++;
            }
        }
        
        return curr;
    }

    //-----PUBLIC-----//
    public int size() {
        return this.length;
    }

    public void inc() {
        this.length++;
    }

    public void dec() {
        this.length--;
    }
 
    //=====SET=====//
    private void setFirst(Node<T> first) {
        this.first = first;
    }

    private void setLast(Node<T> last) {
        this.last = last;
    }

    //=====METODOS=====//
    public Boolean isEmpty() {
        return this.size() == 0;
    }

    //=====ADD=====//
    public void add(T obj) {
        Node<T> node = new Node<>(obj);

        if(isEmpty()) {
            this.setFirst(node);
            this.setLast(node);
        } else {
            this.getLast().setNext(node);
            this.setLast(node);
        }
        inc();
    }

    public void add(T obj, int pos) {
        if(pos < 0 || pos > this.size()) throw new IndexOutOfBoundsException();

        if(pos == 0) insert(obj);
        else if(pos == this.size()) add(obj); 
        else {
            Node<T> curr = node(pos - 1); 
            curr.setNext(new Node<>(obj, curr.getNext()));
            curr = null;
            inc();
        }
    }

    //=====INSERT=====//
    public void insert(T obj) {
        Node<T> node = new Node<>(obj, this.getFirst());
        if(isEmpty()) {
            this.setFirst(node);
            this.setLast(node);
        } else this.setFirst(node);
        node = null;
        inc();
    }

    //=====POP=====//
    public T pop() {
        if(isEmpty()) throw new EmptyLinkedListException();

        T content = this.getLast().getObj();
        if(this.getFirst() == this.getLast()) {
            this.setFirst(null);
            this.setLast(null);
        } else {
            Node<T> curr = this.node(this.size() - 2);
            curr.setNext(null);
            this.setLast(curr);
            curr = null;
        }

        dec();
        return content;
    }

    public T pop(int pos) {
        if(pos < 0 || pos >= this.size()) throw new IndexOutOfBoundsException();
        if(isEmpty()) throw new EmptyLinkedListException();

        T content;
        if(pos == 0) content = del();
        else if(pos == this.size() - 1) content = pop();
        else {
            Node<T> curr = this.node(pos - 1);
            Node<T> rem = curr.getNext();
            content = rem.getObj();
            curr.setNext(rem.getNext());
            rem.setNext(null);
            curr = rem = null;
            dec();
        }

        return content;
    }

    public T pop(T obj) {
        if(isEmpty()) throw new EmptyLinkedListException();

        T content = null;
        int pos = this.search(obj);
        if(pos >= 0) content = pop(pos);
        return content;
    }

    //=====DEL=====//
    public T del() {
        if(isEmpty()) throw new EmptyLinkedListException();

        T content = this.getFirst().getObj();
        if(this.getFirst() == this.getLast()) {
            this.setFirst(null);
            this.setLast(null);
        } else {
            Node<T> curr = this.getFirst();
            this.setFirst(curr.getNext());
            curr.setNext(null);
            curr = null;
        }

        dec();
        return content;
    }

    //=====SEARCH=====//
    public int search(T obj) {
        if(isEmpty()) throw new EmptyLinkedListException();

        int i = 0;
        Node<T> curr = this.getFirst();
        if(obj == null) {
            while(curr != null) {
                if(curr.getObj() == null) {
                    curr = null;
                    return i;
                }
                curr = curr.getNext();
                i++;
            }
        } else {
            while(curr != null) {
                if(curr.getObj().equals(obj)) {
                    curr = null;
                    return i;
                }
                curr = curr.getNext();
                i++;
            }
        }

        return -1;
    }

    //=====CONTAINS=====//
    public Boolean contains(T obj) {
        if(isEmpty()) throw new EmptyLinkedListException();

        Boolean value = false;
        Node<T> curr = this.getFirst();
        if(obj == null) {
            while(!value && curr != null) {
                value = curr.getObj() == null;
                curr = curr.getNext();
            }
        } else {
            while(!value && curr != null) {
                value = curr.getObj().equals(obj);
                curr = curr.getNext();
            }
        }

        curr = null;
        return value;
    }

    //=====REPLACE=====//
    public void replace(T obj, T rep) {
        if(isEmpty()) throw new EmptyLinkedListException();
        int pos = this.search(obj);

        if(pos >= 0) this.node(pos).setObj(rep);
    }

    public void replaceAll(T obj, T rep) {
        if(isEmpty()) throw new EmptyLinkedListException();
        int pos = this.search(obj);

        while(pos >= 0) {
            this.node(pos).setObj(rep);
            pos = this.search(obj);
        }
    }

    //=====OVERRIDE=====//
    @Override 
    public String toString() {
        StringBuffer sb = new StringBuffer("[");
        Node<T> curr = this.getFirst();
        while(curr != null) {
            sb.append(curr.getObj());
            if(curr.getNext() != null) sb.append(", ");
            curr = curr.getNext();
        }
        curr = null;
        sb.append("]");
        return sb.toString();
    }
}

File: EstruturasFlexiveis/ListaSimplesmenteEncadeada/C/LinkedList.c

#include "header/LinkedList.h"

//=====ALOCATION=====//
LinkedList* construct() {
    LinkedList* list = (LinkedList*)malloc(sizeof(LinkedList));
    clear(list);
    return list;
}

void clear(LinkedList* const list) {
    list->first = NULL;
    list->last = NULL;
    list->len = 0;
}

void destruct(LinkedList* const list) {
    for(int i = list->len - 1; i >= 0; i--)     
        free(node(list, i));
    free(list);
}

//=====SEARCH_NODE=====//
Node* node(LinkedList* const list, int idx) {
    Node* curr = list->first;
    
    for(int i = 0; i < idx; i++)
        curr = curr->next;

    return curr;
}

int get(LinkedList* const list, int idx) {
    return node(list, idx)->obj;
}

//=====ADD=====//
void add(LinkedList* const list, int obj) {
    Node* node = constructNodeObj(obj);

    if(list->len == 0) {
        list->first = node;
        list->last = node;
    } else {
        list->last->next = node;
        list->last = node;
    }

    list->len++;
}

void addp(LinkedList* const list, int obj, int idx) {
    if(idx >= 0 && idx <= list->len) {
        if(idx == 0) insert(list, obj);
        else if(idx == list->len) add(list, obj);
        else {
            Node* cell = constructNodeObj(obj);
            Node* curr = node(list, idx - 1);
            cell->next = curr->next;
            curr->next = cell;
            list->len++;
        }
    }
}

//=====INSERT=====//
void insert(LinkedList* const list, int obj) {
    Node* node = constructNodeObj(obj);
    node->next = list->first;

    if(list->len == 0) {
        list->first = node;
        list->last = node;
    } else list->first = node;

    list->len++;
}

//=====POP=====//
int pop(LinkedList* const list) {
    int i = 0;

    if(list->len != 0) {
        if(list->first == list->last) {
            free(list->first);
            list->first = list->last = NULL;
        } else {
            i = list->last->obj;
            list->last = node(list, list->len - 2);
            free(list->last->next);
            list->last->next = NULL;
        }
        list->len--;
    }

    return i;
}

int popp(LinkedList* const list, int idx) {
    int i = 0;

    if(list->len != 0 && idx >= 0 && idx <= list->len - 1) {
        if(idx == 0) i = del(list);
        else if(idx == list->len - 1) i = pop(list);
        else {
            Node* curr = node(list, idx - 1);
            Node* rem = curr->next;
            i = rem->obj;

            curr->next = rem->next;
            rem->next = NULL;
            free(rem);
            
            list->len--;
        }
    } 

    return i;
}

int popObj(LinkedList* const list, int obj) {
    int idx = search(list, obj);

    if(idx >= 0) popp(list, idx);
    return (idx >= 0) ? obj : idx;
}

//=====DEL=====//
int del(LinkedList* const list) {
    int i = 0;

    if(list->len != 0) {
        i = list->first->obj;

        if(list->first == list->last) {
            free(list->first);
            list->first = list->last = NULL;
        } else {
            Node* curr = list->first;
            list->first = curr->next;
            curr->next = NULL;
            free(curr);
        }
        
        list->len--;
    }

    return i;
}

//=====SEARCH=====//
int search(LinkedList* const list, int obj) {
    Node* curr = list->first;
    int i = 0;

    while(curr != NULL) {
        if(curr->obj == obj) return i;
        curr = curr->next;
        i++;
    }

    return -1;
}

//=====CONTAINS=====//
bool contains(LinkedList* const list, int obj) {
    Node* curr = list->first;
    bool value = false;

    while(!value && curr != NULL) {
        value = curr->obj == obj;
        curr = curr->next;
    }

    return value;
}

//=====INVERT=====//
void invert(LinkedList* const list) {
    Node* anchor = list->last;
    Node* curr = NULL;

    while(list->first != anchor) {
        curr = list->first->next;
        list->first->next = anchor->next;
        anchor->next = list->first;
        list->first = curr;
        if(list->last == anchor) list->last = anchor->next;
    }
}

//=====PRINT=====//
void print(LinkedList* const list) {
    Node* curr = list->first;
    printf("[");
    while(curr != NULL) {
        printf("%i", curr->obj);
        if(curr->next != NULL) printf(", ");
        curr = curr->next;
    }
    printf("]\n");
}
